source/dofs/dof_tools_constraints.cc

```diff
@@ -2094,23 +2096,50 @@ namespace DoFTools
             }
 
           // In case of a dependency cycle we, either
-          //  - do nothing if cycle_factor == 1. In this case all degrees
+          //  - do nothing if cycle_constraint_factor == 1. In this case all
+          //  degrees
           //    of freedom are already periodically constrained,
           //  - otherwise, force all dofs to zero (by setting dof_left to
           //    zero). The reasoning behind this is the fact that
-          //    cycle_factor != 1 occurs in situations such as
+          //    cycle_constraint_factor != 1 occurs in situations such as
           //    x1 == x2 and x2 == -1. * x1. This system is only solved by
           //    x_1 = x_2 = 0.
 
           if (constraints_are_cyclic)
             {
-              if (std::abs(cycle_factor - 1.) > eps)
+              if (std::abs(cycle_constraint_factor - 1.) > eps)
                 affine_constraints.add_line(dof_left);
             }
           else
             {
               affine_constraints.add_line(dof_left);
-              affine_constraints.add_entry(dof_left, dof_right, factor);
+              affine_constraints.add_entry(dof_left,
+                                           dof_right,
+                                           constraint_factor);
+              // The number 1e10 in the assert below is arbitrary. If the
+              // absolute value of constraint_factor is too large, then probably
+              // the absolute value of periodicity_factor is too large or too
+              // small. This would be equivalent to an evanescent wave that has
+              // a very small wavelength. A quick calculation shows that if
+              // |periodicity_factor| > 1e10 -> |np.exp(ikd)|> 1e10, therefore k
+              // is imaginary (evanescent wave) and the evanescent wavelength is
+              // 0.27 times smaller than the dimension of the structure,
+              // lambda=((2*pi)/log(1e10))*d. Imaginary wavenumbers can be
+              // interesting in some cases
+              // (https://doi.org/10.1103/PhysRevA.94.033813).In order to
+              // implement the case of in which the wavevector can be imaginary
+              // it would be necessary to rewrite this function and the dof
+              // ordering method should be modified.
+              // Let's take the following constraint a*x1 + b*x2 = 0. You could
+              // just always pick x1 = b/a*x2, but in practice this is not so
+              // stable if a could be a small number -- intended to be zero, but
+              // just very small due to roundoff. Of course, constraining x2 in
+              // terms of x1 has the same problem. So one chooses x1 = b/a*x2 if
+              // |b|<|a|, and x2 = a/b*x1 if |a|<|b|.
+              Assert(
+                std::abs(constraint_factor) < 1e10,
```

Comment:
This code is only valid for real wavenumbers. If the wavenumber is real then the absolute value of `periodic_factor` is one, there is only a phase shift.

In case we want to add support for imaginary wavevectors. I wrote in this comment what it has to be done.

I added an assert in case imaginary wavectors are used.